geoutils/raster/delayed.py

```diff
@@ -0,0 +1,777 @@
+# of chunked grids
+GeoGridType = TypeVar("GeoGridType", bound="GeoGrid")
+
+
+class GeoGrid:
+    """
+    Georeferenced grid class.
+
+    Describes a georeferenced grid through a geotransform (one-sided bounds and resolution), shape and CRS.
+    """
+
+    def __init__(self, transform: rio.transform.Affine, shape: tuple[int, int], crs: rio.crs.CRS | None):
+
+        self._transform = transform
+        self._shape = shape
+        self._crs = crs
+
+    @property
+    def transform(self) -> rio.transform.Affine:
+        return self._transform
+
+    @property
+    def crs(self) -> rio.crs.CRS:
+        return self._crs
+
+    @property
+    def shape(self) -> tuple[int, int]:
+        return self._shape
+
+    @property
+    def height(self) -> int:
+        return self.shape[0]
+
+    @property
+    def width(self) -> int:
+        return self.shape[1]
+
+    @property
+    def res(self) -> tuple[int, int]:
+        return self.transform[0], abs(self.transform[4])
+
+    @property
+    def bounds(self) -> rio.coords.BoundingBox:
+        return rio.coords.BoundingBox(*rio.transform.array_bounds(self.height, self.width, self.transform))
+
+    def bounds_projected(self, crs: rio.crs.CRS = None) -> rio.coords.BoundingBox:
+        if crs is None:
+            crs = self.crs
+        return _get_bounds_projected(bounds=self.bounds, in_crs=self.crs, out_crs=crs)
+
+    @property
+    def footprint(self) -> gpd.GeoDataFrame:
+        return _get_footprint_projected(self.bounds, in_crs=self.crs, out_crs=self.crs, densify_points=100)
+
+    def footprint_projected(self, crs: rio.crs.CRS = None) -> gpd.GeoDataFrame:
+        if crs is None:
+            crs = self.crs
+        return _get_footprint_projected(self.bounds, in_crs=self.crs, out_crs=crs, densify_points=100)
+
+    @classmethod
+    def from_dict(cls: type[GeoGridType], dict_meta: dict[str, Any]) -> GeoGridType:
+        """Create a GeoGrid from a dictionary containing transform, shape and CRS."""
+        return cls(**dict_meta)
+
+    def shift(
+        self: GeoGridType,
+        xoff: float,
+        yoff: float,
+        distance_unit: Literal["georeferenced"] | Literal["pixel"] = "pixel",
+    ) -> GeoGridType:
+        """Shift into a new geogrid (not inplace)."""
+
+        if distance_unit not in ["georeferenced", "pixel"]:
+            raise ValueError("Argument 'distance_unit' should be either 'pixel' or 'georeferenced'.")
+
+        # Get transform
+        dx, b, xmin, d, dy, ymax = list(self.transform)[:6]
+
+        # Convert pixel offsets to georeferenced units
+        if distance_unit == "pixel":
+            # Multiplying the offset by the resolution might amplify floating precision issues
+            # xoff *= self.res[0]
+            # yoff *= self.res[1]
+
+            # Using the boundaries instead!
+            xoff = xoff / self.shape[1] * (self.bounds.right - self.bounds.left)
```

Comment:
How do you know this will have less floating precision issues? Also, maybe first multiplying by the image range then dividing might cause less issues?

Reply:
It might not matter at all, this is a relic from when I thought the Rasterio issue came from this code. Changed the comment.